sdk/python/feast/infra/offline_stores/contrib/mssql_offline_store/tests/data_source.py

```diff
@@ -0,0 +1,112 @@
+from typing import Dict, List
+
+import pandas as pd
+import pytest
+from sqlalchemy import create_engine
+from testcontainers.core.container import DockerContainer
+from testcontainers.core.waiting_utils import wait_for_logs
+from testcontainers.mssql import SqlServerContainer
+
+from feast.data_source import DataSource
+from feast.infra.offline_stores.contrib.mssql_offline_store.mssql import (
+    MsSqlServerOfflineStoreConfig,
+    _df_to_create_table_sql,
+)
+from feast.infra.offline_stores.contrib.mssql_offline_store.mssqlserver_source import (
+    MsSqlServerSource,
+)
+from feast.saved_dataset import SavedDatasetStorage
+from tests.integration.feature_repos.universal.data_source_creator import (
+    DataSourceCreator,
+)
+
+MSSQL_USER = "SA"
+MSSQL_PASSWORD = "yourStrong(!)Password"
+
+
+# This is the sql container to use if your machine doesn't support the official msql docker container.
+@pytest.fixture(scope="session")
+def mssql_container():
+    container = (
+        DockerContainer("mcr.microsoft.com/azure-sql-edge:1.0.6")
+        .with_exposed_ports("1433")
+        .with_env("ACCEPT_EULA", "1")
+        .with_env("MSSQL_USER", MSSQL_USER)
+        .with_env("MSSQL_SA_PASSWORD", MSSQL_PASSWORD)
+    )
+    container.start()
+    log_string_to_wait_for = "Service Broker manager has started"
+    wait_for_logs(container=container, predicate=log_string_to_wait_for, timeout=30)
+
+    yield container
+    container.stop()
+
+
+class MsSqlDataSourceCreator(DataSourceCreator):
+    tables: List[str] = []
+
+    def __init__(
+        self, project_name: str, fixture_request: pytest.FixtureRequest, **kwargs
+    ):
+        super().__init__(project_name)
+        self.tables_created: List[str] = []
+        self.container = SqlServerContainer(user=MSSQL_USER, password=MSSQL_PASSWORD)
+        self.container.start()
+        if not self.container:
+            raise RuntimeError(
+                "In order to use this data source "
+                "'feast.infra.offline_stores.contrib.mssql_offline_store.tests' "
+                "must be include into pytest plugins"
+            )
```

Comment:
you likely need to use the wait_for_logs call above as well here btw since otherwise, the test might start before the container is ready